### ISSUE TITLE
FPS drop in 3.2.4 RC 3 when printing Control._set_size warnings

### ISSUE THREAD

**Issue by Error7Studios** [godotengine/godot#46376]
**Godot version:**
`3.2.4 RC 3`

**OS/device including version:**
Windows 10

**Issue description:**
Bad FPS drop in `3.2.4 RC 3`. (Okay in `RC 2`.)
Also, getting hundreds of these warnings:
`W 0:00:01.415   _set_size: Adjusting the size of Control nodes before they are fully initialized is unreliable. Consider deferring it with set_deferred().`
(Don't know if related to FPS drop or not)
I don't know how to disable/ignore this warning. (I don't want to use `set_deferred()`)

**Steps to reproduce:**
Run the attached project.
Move the character as shown in [the example video](https://user-images.githubusercontent.com/60025044/109005975-27bc5200-7670-11eb-90e9-a6bf359f92b3.mp4).
Select `End` from the menu to continue moving character
```
Controls:
- Arrow keys: Move cursor
- Enter: Select menu option / Select cell to move character to
- Escape: Quit
```

**Minimal reproduction project:**
[FPS Drop in RC 3.zip](https://github.com/godotengine/godot/files/6036218/FPS.Drop.in.RC.3.zip)

**Comment by Calinou** [comment]
> (I don't want to use `set_deferred()`)

Why not?

**Comment by Error7Studios** [comment]
I tried it, and it totally screws up my menu GUI. The menu is sized based on the child `BTN` nodes that compose it. 
(See `create(...)` in `Menu.gd`)
I can try putting some yields to make sure it sizes itself correctly, but I haven't tried it yet.
I try to avoid `yield` whenever possible.

**Comment by akien-mga** [comment]
Tested on Linux with Intel HD Graphics 630, I get:
- ~800 FPS in 3.2.4 RC 2
- ~400-500 FPS in 3.2.4 RC 3

I'm not sure it's related to the warning spam though as this only happens on the first frame, there are no warnings once the scene has been initialized, so the actual FPS drop might be related to rendering changes (CC @lawnjelly @clayjohn).

As for the Control warning, CC @KoBeWi @groud to assess if this MRP is a valid use case where the warning is a bother, or if OP should do things differently to avoid relying on potentially undefined behavior.

**Comment by KoBeWi** [comment]
I checked #24859 and @groud mentions in one comment that setting size won't work with some specific anchor settings. Maybe we could rework the warning to only show when it might actually fail. I remember seeing it in my project too.

**Comment by Error7Studios** [comment]
> the actual FPS drop might be related to rendering changes

In that case, here's my graphics info.
I don't know what you need to know, so here's everything:

```
OpenGL ES 3.0 Renderer: GeForce GTX 1050 Ti/PCIe/SSE2
OpenGL ES Batching: ON
```

```
Graphics
		Monitor
			Name	BenQ GL2760 on NVIDIA GeForce GTX 1050 Ti
			Current Resolution	1920x1080 pixels
			Work Resolution	1920x1080 pixels
			State	Enabled, Primary
			Monitor Width	1920
			Monitor Height	1080
			Monitor BPP	32 bits per pixel
			Monitor Frequency	60 Hz
			Device	\\.\DISPLAY1\Monitor0
		Intel HD Graphics 630
			Manufacturer	Intel
			Model	HD Graphics 630
			Device ID	8086-5912
			Revision	5
			Subvendor	Lenovo (17AA)
			Current Performance Level	Level 0
			Current GPU Clock	0 MHz
			Voltage	0.675 V
			Driver version	26.20.100.7986
				Count of performance levels : 1
					Level 1 - "Perf Level 0"
		NVIDIA GeForce GTX 1050 Ti
			Manufacturer	NVIDIA
			Model	GeForce GTX 1050 Ti
			Device ID	10DE-1C82
			Revision	A2
			Subvendor	EVGA (3842)
			Current Performance Level	Level 0
			Current GPU Clock	139 MHz
			Current Memory Clock	405 MHz
			Current Shader Clock	405 MHz
			Voltage	0.675 V
			Technology	14 nm
			Bus Interface	PCI Express x16
			Temperature	33 °C
			Driver version	27.21.14.5671
			BIOS Version	86.07.39.00.52
			Memory	4095 MB
				Count of performance levels : 1
						Level 1 - "Perf Level 0"
							GPU Clock	139 MHz
							Shader Clock	405 MHz
```

**Comment by lawnjelly** [comment]
I'm not seeing much if any drop in fps between RC2 and RC3 on Linux Mint. I don't think anything much has changed in 2d rendering. The only thing I would watch for at the moment if there have been any side effects to the changes to atomics. That could easily cause performance changes.

**Comment by lawnjelly** [comment]
See also #41513 which seems to confirm the changes to atomics hypothesis.

**Comment by akien-mga** [comment]
> See also #41513 which seems to confirm the changes to atomics hypothesis.

I tried the CI builds from around the atomics changes and I get ~1800 FPS with all of the commits I tried, so this might be a red herring for this specific MRP. At a glance I don't see what later commit could cause this. I'll do a proper bisect.

**Comment by akien-mga** [comment]
Welp I lost some time bisecting to reach an unrelated commit: 92687b4544. The problem is that this commit makes Godot use my discrete AMD GPU instead of the integrated Intel one which was used previously, and strangely enough the performance is worse with the AMD GPU for this test case. That's the risk of comparing good performance with better performance (400 FPS vs 1800 FPS).

I'll restart bisecting with `DRI_PRIME=0` to force using the Intel IGP.

**Comment by akien-mga** [comment]
So in the end I get the same performance in 3.2.4 RC 2 and RC 3 when using `DRI_PRIME=0` to force using the Intel IGP.
So I can't confirm the regression anymore. Interestingly, I have better (measured) performance in local `debug` builds than the official `release_debug` builds (1100 FPS vs 800 FPS).

@Error7Studios What kind of FPS drop are you seeing exactly?

**Comment by akien-mga** [comment]
Alright I finally noticed that there are more detailed steps to reproduce and a video.

The stutter shown in the video does come from printing 100 warnings in one frame about setting the Control size in `_ready`, so my debugging of potential rendering issues was mislead by the Intel/AMD change for my config.

So reverting or improving #42427 should indeed solve this.

**Comment by RandomShaper** [comment]
I can also confirm that https://github.com/godotengine/godot/commit/a8105d73c7eac8998a90bb02c10aeb0da6002a76 is the "culprit". I mean, if the message is meaningful and for debug builds only, I think it's fine and can maybe be improved via one of those macros that print only once for the same error.

The other FPS drop is surely related to the MT work, but I don't think that's the one the OP is complaining about. I'll keep investigating it.

**Comment by Error7Studios** [comment]
The FPS drop I was talking about is when moving the character to a new position on the map.
When you do so, a menu gets opened (created), and that's when all the Control resizing happens, and I get like 200 warnings all at once.
In RC3, it drops down to around 5 FPS when this happens.
In RC2, I don't get the warnings, and there's no FPS drop when the menu opens.
Both run at 60 FPS when idle, though.
(Was about to try to uncap the FPS and see the difference, but Windows decided to update when I tried to update my NVidea driver)

**Comment by RandomShaper** [comment]
For the record, on Windows 10 these are my numbers (using the official builds):
~2600-2800 FPS in 3.2.4 RC 2 and 3
In RC 3 it seems it's biasing towards 2800, but that may just be me. :)

**Comment by akien-mga** [comment]
> The other FPS drop is surely related to the MT work, but I don't think that's the one the OP is complaining about. I'll keep investigating it.

The FPS drop that I reported initially is not related to MT, just to different priorities for Intel and AMD GPUs, see https://github.com/godotengine/godot/issues/46376#issuecomment-785838134
When I force selecting a given GPU I do get the same performance in both RC 2 and RC 3.

**Comment by RandomShaper** [comment]
> (Was about to try to uncap the FPS and see the difference, but Windows decided to update when I tried to update my NVidia driver)

If you finally do, I'd expect you to the idle uncapped FPS to be similar, and to drop a lot as well during the prints. Windows console is slow and many prints block the game for a big amount of time. You could also try to disable stdout and stderr in project settings (that will also prevent the FPS prints if enabled, but at least you'll be able to notice in the game itself that the drop doesn't happen).

**Comment by akien-mga** [comment]
I reverted the cherry-pick of #42427 as part of #46417, so this issue should be fixed in the 3.2.4 RC 4.

**Comment by Error7Studios** [comment]
> If you finally do, I'd expect your idle uncapped FPS to be similar between both RCs, and to drop a lot as well during the prints in RC 3

I just tried this, and yep, getting ~2800 idle FPS in RC2 & 3.
It dropped down to a low of 103 FPS during the prints, but was actually completely frozen for almost 2 seconds.

> You could also try to disable stdout and stderr in project settings

Also just tried this, and you're right, the FPS drop is gone!

> I reverted the cherry-pick of #42427 as part of #46417, so this issue should be fixed in the 3.2.4 RC 4.

You guys are absolute gods. Thank you for all your hard work.
